Possibility to switch between FIPS and non-FIPS during keycloak+quarkus seerver build

mposolda opened the pull request:
This prototype PR handles:

- Build Keycloak quarkus distribution either with the plain `bouncycastle` dependencies (same behaviour as current Keycloak) or with `bouncycastle-fips` dependencies

- PR introduces the `SecurityProvider` interface to handle differences between bouncycastle APIs. As sometimes the code and API calls may need to be different according to if you use `bouncycastle` or `bouncycastle-fips` dependencies. The default implementation is still use the  code dependent on the `bouncycastle` . So this PR should not break anything (hopefully) as the default implementation uses same behaviour as current Keycloak upstream (so non-fips by default).

- PR introduces `keycloak-fips-integration` module with the implementation of `SecurityProvider` using `bouncycastle-fips` . So the idea is, that for all the codes, where the differences are between the calls to `bouncycastle` and `bouncycastle-fips`, the new method will be introduced to the `SecurityProvider` . The module `keycloak-fips-integration` has dependencies on the `bouncycastle-fips` libraries.

- The maven profile is introduced in the `quarkus` module to allow building with FIPS. By default, project is built with non-fips based profile, however there is possibility to build with the `fips` profile, which will add the `bouncycastle-fips` dependencies and the new module `keycloak-fips-integration` to the keycloak + quarkus. The instructions on how to build keycloak+quarkus with bouncycastle-fips is described in the README inside docs/fips.md

mposolda committed:
Possibility to switch between FIPS and non-FIPS during keycloak+quarkus server build

Closes #12522

stianst commented:
LGTM as an initial approach.

Is SecurityProvider the best name? Security is a bit broad term. Perhaps CryptoProvider would be clearer?

Not sure integration is the best place, as that's mostly client side stuff today. Would it make sense to add a new parent module "crypto" with two child modules "default" and "fips". Then make sure default and fips has the correct dependencies, which can then be used to include the BouncyCastle libraries elsewhere. This would allow not only the Quarkus dist itself to have a profile for default/fips, but also other modules, which would in term allow us to verify at compile time with bc-fips and not just at runtime, if we have a global "default" vs "fips" profile. Perhaps also the "fips" module should be named "fips 140-2" as I'd imagine we'll end up with a different profile again for 140-3.

I can also see that SecurityProvider could expand further as we're moving ahead with this work. For example it can be the thing that is used to check what "crypto mode" Keycloak is in, what algorithms are supported, etc. when we come to wanting to enable/disable stuff in admin console.

stianst commented:
> Do you mean that module `keycloak-core` won't have any dependency on bouncycastle at all? I agree with that, but I guess it would be quite a lot of changes and not sure whether to do it all in this PR? It seems it might be better to iteratively move individual changes to `CryptoProvider` implementation and then at some point move the "crypto / bouncycastle" stuff to the `crypto/default` module. Also it will require some changes in the packaging and distribution (including adapters, wildfly modules etc) as all the parties will need this new `crypto-default` module.
> 
> So for now, I've introduced only module `crypto/fips1402` (moved the module from `integration` and added `1402` suffix). Is it ok for this PR and then we will continue to evolve in the follow-up PR in this direction?
> 

Yeah, basically what I was thinking was that no modules have direct deps on BC, but instead gets these through either crypto-default or crypto-fips1402, then we'd have profiles to allow switching at build/compile time so we can check the API issues at compile time.

I agree it's not something to address in this PR, but rather a follow-up, if at all. Not sure the effort of doing it is justified by the benefits by being able to check APIs at build/compile time. I think it is though.

stianst commented on core/src/main/java/org/keycloak/crypto/integration/CryptoIntegration.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class CryptoIntegration {
```
Should probably consider getting rid of BouncyIntegration and merging that stuff into here. Can we done as a follow-up perhaps?

stianst commented on core/src/main/java/org/keycloak/crypto/integration/CryptoIntegration.java:
```diff
@@ -0,0 +1,47 @@
+import org.keycloak.common.util.BouncyIntegration;
+import org.keycloak.common.util.reflections.Reflections;
+
+/**
```
Ideally this should be a KC SPI, and not something different, but for the time being we need this to work with adapters as well, but would be good to refactor this when we no longer have any adapters.

mposolda replied:
Yes, exactly. I've considered Keycloak SPI from the beginning, but not possible due it needs to work with adapters as well.

stianst commented on core/src/main/java/org/keycloak/crypto/integration/CryptoIntegration.java:
```diff
@@ -0,0 +1,47 @@
+
+    // Try to auto-detect provider. Currently autodetected based on whether we have BC or BCFIPS on the classpath. Should be probably improved and rather
+    // detected based on the system security settings or with ServiceLoader etc. Ideally, the BouncyCastle implementation should be chosen based on the SecurityProvider rather than vice-versa.
+    private static CryptoProvider detectProvider() {
```
Maybe it would be better to use Java Service Providers instead of reflection?

stianst commented on pom.xml:
```diff
@@ -75,6 +75,12 @@
         <bouncycastle.version>1.68</bouncycastle.version>
+
+        <!-- TODO Are these correct versions? -->
+        <bouncycastle.fips.version>1.0.2</bouncycastle.fips.version>
+        <bouncycastle.pkixfips.version>1.0.3</bouncycastle.pkixfips.version>
+        <bouncycastle.tlsfips.version>1.0.11</bouncycastle.tlsfips.version>
```
I used some newer versions here:

https://github.com/stianst/keycloak/blob/24f10231fd49db6c34fcca222d720b683f523069/quarkus/runtime/pom.xml#L526

mposolda replied:
Updated per your versions. I guess BouncyCastle provides some mapping to compare BouncyCastle FIPS versions to BouncyCastle non-fips version. I wonder if we should investigate this further...

stianst commented on quarkus/runtime/pom.xml:
```diff
@@ -583,6 +563,88 @@
+                </property>
+            </activation>
+            <dependencies>
+                <dependency>
+                    <groupId>org.bouncycastle</groupId>
+                    <artifactId>bcpkix-jdk15on</artifactId>
+                    <exclusions>
+                        <exclusion>
+                            <groupId>*</groupId>
+                            <artifactId>*</artifactId>
+                        </exclusion>
+                    </exclusions>
+                </dependency>
+                <dependency>
+                    <groupId>org.bouncycastle</groupId>
+                    <artifactId>bcprov-jdk15on</artifactId>
+                    <exclusions>
+                        <exclusion>
+                            <groupId>*</groupId>
+                            <artifactId>*</artifactId>
+                        </exclusion>
+                    </exclusions>
+                </dependency>
```
Can we inherit these dependencies from crypto-default and crypto-fips1402 instead of listing them again here?

mposolda committed:
Update according to the feedback

mposolda commented:
@stianst Addressed your comments and added https://github.com/keycloak/keycloak/issues/12608 as a follow-up. Could you please re-review? I've used separate commit for easier review the latest changes, but I've plan to squash during merge.

stianst commented on core/src/main/java/org/keycloak/crypto/integration/DefaultCryptoProvider.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class DefaultCryptoProvider implements CryptoProvider {
```
Is the plan to move this to crypto/default later? Or should it remain here permanently?

mposolda replied:
Yes, the plan is to do it later. Moving it now will mean that module "crypto/default" will be new mandatory module for the default distribution (including adapters), which will mean lots of additional changes in this PR in the packaging etc.

IMO The ideal final state will be, that modules like keycloak-common or keycloak-core don't have dependency on bouncyCastle and hence the crypto stuff is incrementally moved into "crypto/default" and "crypto/fips" and at this stage the DefaultCryptoProvider will be in "crypto/default" .

stianst commented:
Approved, can you create an issue for moving DefaultCryptoProvider to its own module?

mposolda commented:
@stianst Thanks! Added https://github.com/keycloak/keycloak/issues/12625
